File: packages/dart/test/contract_test.dart
// 契约测试：以根 tests/ 的 Schema + Fixture 验证 Dart 模型。
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:quanttide_course/quanttide_course.dart';

final fixturesDir = Directory('${Directory.current.path}/../../tests/fixtures');
final schemasDir = Directory('${Directory.current.path}/../../tests/schemas');

Map<String, dynamic> fixture(String name) =>
    jsonDecode(File('${fixturesDir.path}/$name').readAsStringSync())
        as Map<String, dynamic>;

void hasRequired(Map<String, dynamic> data, String schemaName) {
  final schema = jsonDecode(
    File('${schemasDir.path}/$schemaName.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final missing =
      ((schema['required'] as List?) ?? []).cast<String>().where((k) => !data.containsKey(k));
  expect(missing, isEmpty, reason: '$schemaName 缺少必填字段');
}

void main() {
  test('course 契约', () {
    final data = fixture('course.json');
    hasRequired(data, 'course');
    final c = Course.fromJson(data);
    expect(c.id, 'prod');
    expect(c.slug, 'prod');
    expect(c.status, 'published');
  });

  test('lesson 契约 + round-trip', () {
    final data = fixture('lesson.json');
    hasRequired(data, 'lesson');
    final l = Lesson.fromJson(data);
    expect(l.courseId, 'prod');
    expect(l.startSceneId, 'scen-1');
    expect(l.criteria.length, 2);

    // Round-trip：序列化回 API 形态后可再次解析，值不变
    final again = Lesson.fromJson(l.toJson());
    expect(again.id, l.id);
    expect(again.courseId, l.courseId);
    expect(again.startSceneId, l.startSceneId);
    expect(again.criteria, l.criteria);
  });

  test('scene 契约', () {
    final data = fixture('scene.json');
    hasRequired(data, 'scene');
    final s = Scene.fromJson(data);
    expect(s.videoUrl.endsWith('open.mp4'), isTrue);
    expect(s.steps.length, 1);
    expect(s.choices, isEmpty); // 空数组 = 终结场景（字段始终存在）
  });

  test('criterion 契约', () {
    final data = fixture('criterion.json');
    hasRequired(data, 'criterion');
    final cr = Criterion.fromJson(data);
    expect(cr.sceneId, 'scen-1');
    expect(cr.title, '会连接 Zed');

    // 无 sceneId 表示课时级标准
    final lessonLevel = const Criterion(
      id: 'cri-2',
      lessonId: 'less-1',
      title: '完成全部场景',
      description: '课时内所有场景通过',
    );
    expect(lessonLevel.sceneId, isNull);
  });
}

File: packages/go/pkg/contract_test.go
// 契约测试：以根 tests/ 的 Schema + Fixture 为单一事实源验证 Go 模型。
//
// 当前为轻量实现（零依赖）：校验必填字段存在性（与 Schema required 一致）
// 并反序列化标本。完整 JSON Schema 校验待引入校验依赖后升级。
package course_test

import (
	"encoding/json"
	"os"
	"testing"

	course "github.com/quanttide/quanttide-course-toolkit/packages/go/pkg"
)

func fixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile("../../../tests/fixtures/" + name)
	if err != nil {
		t.Fatalf("read %s: %v", name, err)
	}
	return data
}

// hasFields 校验解码后的对象包含 schema 要求的必填字段。
func hasFields(t *testing.T, data []byte, required ...string) map[string]any {
	t.Helper()
	var m map[string]any
	if err := json.Unmarshal(data, &m); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, k := range required {
		if _, ok := m[k]; !ok {
			t.Fatalf("%s 缺少必填字段 %q", "fixture", k)
		}
	}
	return m
}

func TestContract_Fixtures_Unmarshal(t *testing.T) {
	var c course.Course
	if err := json.Unmarshal(fixture(t, "course.json"), &c); err != nil {
		t.Fatal(err)
	}
	hasFields(t, fixture(t, "course.json"), "id", "name", "slug")
	if c.ID != "prod" || c.Status != "published" || c.SortOrder != 1 {
		t.Fatalf("course = %+v", c)
	}

	var l course.Lesson
	if err := json.Unmarshal(fixture(t, "lesson.json"), &l); err != nil {
		t.Fatal(err)
	}
	hasFields(t, fixture(t, "lesson.json"), "id", "courseId", "title", "slug")
	if l.CourseID != "prod" || l.StartSceneID != "scen-1" || len(l.Criteria) != 2 {
		t.Fatalf("lesson = %+v", l)
	}

	var s course.Scene
	if err := json.Unmarshal(fixture(t, "scene.json"), &s); err != nil {
		t.Fatal(err)
	}
	hasFields(t, fixture(t, "scene.json"), "id", "lessonId", "slug", "videoUrl", "choices")
	if s.VideoURL == "" || len(s.Steps) != 1 || s.Choices == nil {
		t.Fatalf("scene = %+v", s)
	}

	var cr course.Criterion
	if err := json.Unmarshal(fixture(t, "criterion.json"), &cr); err != nil {
		t.Fatal(err)
	}
	hasFields(t, fixture(t, "criterion.json"), "id", "lessonId", "title", "description")
	if cr.SceneID != "scen-1" || cr.Title != "会连接 Zed" {
		t.Fatalf("criterion = %+v", cr)
	}
}

// TestContract_RoundTrip 反序列化 → 再序列化，契约字段保持一致。
func TestContract_RoundTrip(t *testing.T) {
	var l course.Lesson
	if err := json.Unmarshal(fixture(t, "lesson.json"), &l); err != nil {
		t.Fatal(err)
	}
	out, err := json.Marshal(l)
	if err != nil {
		t.Fatal(err)
	}
	var back course.Lesson
	if err := json.Unmarshal(out, &back); err != nil {
		t.Fatal(err)
	}
	if len(back.Criteria) != len(l.Criteria) || back.ID != l.ID || back.CourseID != l.CourseID ||
		back.StartSceneID != l.StartSceneID || back.Duration != l.Duration {
		t.Fatalf("round-trip mismatch: got %#v want %#v", back, l)
	}
	for i := range l.Criteria {
		if back.Criteria[i] != l.Criteria[i] {
			t.Fatalf("criteria[%d] mismatch", i)
		}
	}
}
